src/librustc_builtin_macros/asm.rs

```diff
@@ -147,8 +147,27 @@ fn parse_args<'a>(
                 }
             }
             ast::InlineAsmOperand::Sym { expr }
+        } else if allow_templates {
+            let template = p.parse_expr()?;
+            // If it can't possibly expand to a string, provide diagnostics here to include other
+            // things it could have been.
+            match template.kind {
+                ast::ExprKind::Lit(ast::Lit { kind: ast::LitKind::Str(..), .. }) => {}
+                ast::ExprKind::MacCall(..) => {}
+                _ => {
+                    let errstr = "expected operand, options, or additional template string";
```

Comment:
I'm not too happy about this error message since unlike the one from `expect` it doesn't give you a list of all the available operand types for asm (`in`, `out`, `sym`, etc). However I don't think we can do much better here anyways, so it's probably fine as it is.

Reply:
I originally did copy the full list into this error string; I just changed it because it seemed unwieldy. I'm happy to put it back if you'd prefer. It would look like this:

```
error: expected `const`, `in`, `inlateout`, `inout`, `lateout`, `options`, `out`, `sym`, or additional template string
```

Or perhaps this:

```
error: expected additional template string, operand (`const`, `in`, `inlateout`, `inout`, `lateout`, `out`, `sym`), or `options`
```

Does that seem better?

Reply:
I think it's fine as it is, I can't really think of anything better than what you already have.

Reply:
Alright, I'll leave it as-is.